### ISSUE TITLE
iostat for spdk devices. ?

### ISSUE THREAD

**Issue by alokkataria** [sysstat/sysstat#257]
Hey,

For nvme devices that are managed by user space drivers like spdk, we don't get a block device exported under sysfs nor do the stats show up in /proc/diskstats as there's no kernel backing for these devices. As a result, iostat doesn't expose stats for those devices, SPDK subsystem or the applications built on top of them can easily expose these stats in the format that the kernel exposes them in /proc/diskstats, the only change needed is to ensure iostat also reads stats from those files in addition to reading the proc file. 

Are there any plans to do support these ? Can you take a look at the attached patch which does that and let me know if you would be willing to merge something like this ?

[iostat-spdk-patch.txt](https://github.com/sysstat/sysstat/files/4475977/iostat-spdk-patch.txt)

**Comment by sysstat** [comment]
I'm not familiar with NVMe devices nor with SPDK subsystem. Yet if statistics for such devices may be available only through an alternate stat file, then yes, it would be cool to add support for this to iostat.
Maybe instead of using an environment variable we could consider adding an option like `iostat -f /my/alternate/statfile` to tell iostat to read statistics from another file than `/proc/diskstats`.
You can send me a patch/PR and I will accept it.
I will then maybe update it/add more features, like the possibility to specify a directory name (instead of a file name) with option -f, which would then be considered as something equivalent to `/sys/block` directory containing stats for individual devices and partitions. I imagine that if the userspace driver can expose stats in a `/proc/diskstats`-like file, it would also be feasible to expose them in separate individual files...
The idea is to not always display the stats for all the devices in the system (using -p ALL) just to get stats for some NVMe devices.
Any comments?

**Comment by alokkataria** [comment]
Hi Sebastian,

I like that idea to make it work without "-p ALL" as well, just to make sure I understand your suggestion. I can add an option -f 
which specifies /my/alternate/directory/

If SPDK is driving NVME devices with device names as sda/sdb, that directory will have the following files -

diskstats
sda/stats
sdb/stats
..

If that sounds fine, I can post a patch in my morning tomorrow, do you just want me to upload it here ?

**Comment by sysstat** [comment]
Yes that sounds fine: Option -f will then always point at a directory (and not directly at a plain file).
This directory will contain either a `diskstats` file (which will be used when option `-p ALL` is entered on the command line, the same way `/proc/diskstats` is used without option -f), or files for individual devices: `sda/stat`, `sdb/stat`... and maybe even for partitions: `sda/sda1/stat`, `sda/sda2/stat`... if existent. Almost all the code is already in iostat.
The only limitation I see is that you will be able to display statistics for either standard devices (using `/proc/diskstats` and `/sys/block/*` files) **or** NVMe devices (using your alternate directory location specified with option -f). If this doesn't sound acceptable then maybe an additional option should be considered to tell iostat to use both directory trees...

**Comment by sysstat** [comment]
Oh, and for your patch, the preferred way is to make a pull request.
But if you just want to upload it here, don't worry I will take it anyway :)

**Comment by jimharris** [comment]
Thanks @alokkataria - upstream SPDK could definitely add support to generate iostat-compliant stat files.

Could the patch allow multiple -f options to be specified?  Then if user wants to show data for both kernel and SPDK block devices, they could do something like -f /proc -f /tmp/spdk.

**Comment by alokkataria** [comment]
So I am thinking if -p ALL is provided we read both the /proc/diskstats as well as custom path. That ways we can print data for all devices, irrespective of whether those are kernel managed or spdk managed. 
My existing patch does exactly that but with an environment variable instead of "-f" option.

**Comment by jimharris** [comment]
That works for me.

**Comment by alokkataria** [comment]
Hi Sebastian, 

Attached is the change which adds the -f option and updates the code for the -p ALL case. I didn't get time today to get thru the individual devices cases today, I need to think thru the /sys/dev/block/major:minor/stats case that is handled in read_sysfs_part_stat. 
Besides, I am thinking read_sysfs_all_devices_stat maybe should be updated to cycle thru both the /sys/block as well as the "additional_dir" paths.
In anycase, I think we can fix these rest of the issues in a separate change maybe ?

Thanks,
Alok
[0001-Extend-IOStat-to-expose-stats-for-spdk-managed-device-patch.txt](https://github.com/sysstat/sysstat/files/4487701/0001-Extend-IOStat-to-expose-stats-for-spdk-managed-device-patch.txt)

**Comment by sysstat** [comment]
Alok,
Thanks for your patch. I'm rather busy with other changes in sysstat at the present time but I will merge it ASAP.

**Comment by sysstat** [comment]
Hi,
Commit e1b85a9 adds support for devices that are managed in userspace like spdk.
From the commit message:

Add new flag to `iostat` that can be used to specify an alternate directory for devices statistics.

Example:

`iostat -f /altdir [...]`
-> Use `<altdir>` directory to read devices statistics.

`iostat +f /altdir [...]`
-> Use standard kernel files **and** `<altdir>` to read devices statistics.

`<altdir>` is a directory containing files with statistics for devices managed in userspace.

`<altdir>` may contain:
- a `diskstats` file whose format is compliant with that located in `/proc`.
- statistics for individual devices contained in files whose format is compliant with that of files located in `/sys`.

In particular, the following files located in `<altdir>` may be used by `iostat`:
```
<altdir>/block/<device>/stat
<altdir>/block/<device>/<partition>/stat
```
`<partition>` files must have an entry in `<altdir>/dev/block/` directory,
e.g.:
```
<altdir>/dev/block/[major]:[minor] --> ../../block/<device>/<partition>
```
Notes:
1) `iostat` uses the `/proc/diskstats` file to read statistics only when "-p all" has been entered on the command line (read statistics for all the devices and their partitions).
2) `iostat` uses the `/sys/block/<device>/stat` files to read the statistics for a device (e.g. `iostat sda`) and possibly all its partitions (e.g. `iostat -p sda`).
3) `iostat` uses the link in `/sys/dev/block/[major]:[minor]` to know where the `stat` file is located for a partition that has been entered on the command line (e.g. `iostat sda3`). The partition must exist in `/dev` to get its major and minor numbers.

Please test and tell me if anywthing goes wrong.
